File: problems/0001-two-sum/python/solution.py
# Time: O(n) | Space: O(n)
# Hash map stores each number and its index while we scan once.


def two_sum(nums: list[int], target: int) -> list[int]:
    seen: dict[int, int] = {}

    for index, value in enumerate(nums):
        needed = target - value
        if needed in seen:
            return [seen[needed], index]
        seen[value] = index

    raise ValueError("No valid pair exists")

File: problems/0001-two-sum/javascript/solution.js
// Time: O(n) | Space: O(n)
// Store each seen number and its index, then look up the complement.

function twoSum(nums, target) {
  const seen = new Map();

  for (let index = 0; index < nums.length; index += 1) {
    const needed = target - nums[index];
    if (seen.has(needed)) {
      return [seen.get(needed), index];
    }
    seen.set(nums[index], index);
  }

  throw new Error("No valid pair exists");
}

module.exports = { twoSum };
